### PR TITLE
Respect options from baseline

### PR DESCRIPTION
**Pull request by killuazhu**
Fully implements the proposal in https://github.com/Yelp/detect-secrets/issues/121#issuecomment-459575817. More than half of the changes are new test cases.

Some logic can be implemented in different places, I chose the current way based on my understanding. Let me know if you have any suggetsions or comments.

FYI @KevinHock @domanchi
CC @jribm

### CODE REVIEW

File: detect_secrets/main.py [hunk 1]
```diff
@@ -39,10 +40,12 @@ def main(argv=None):
...
 
         else:
-            baseline_dict = _perform_scan(
-                args,
-                plugins,
-            )
+            if args.import_filename:
+                plugins = initialize.merge_plugin_from_baseline(
+                    _get_plugin_from_baseline(args.import_filename), args,
+                )
```
Comment (domanchi) [inline detect_secrets/main.py]: Let's move this block inside `_perform_scan`, right after we get the old_baseline.

e.g.

```
def _perform_scan(args, plugins):
    old_baseline = _get_existing_baseline(args.import_filename)
    if old_baseline:
        plugins = initialize.merge_plugin_from_baseline(...)
```

This way, we don't need to incur two file reads. Also, I'm not entirely sure that reading from `stdin` would work twice.

File: tests/main_test.py [hunk 2]
```diff
@@ -163,10 +164,186 @@ def test_old_baseline_ignored_with_update_flag(
...
+            assert main(
+                shlex.split(
+                    'scan --update old_baseline_file {}'.format(
+                        plugins_overwriten,
+                    ),
+                ),
+            ) == 0
+
+            print("Used:", file_writer.call_args[1]['data']['plugins_used'])
+            print("Wrote:", plugins_wrote)
```
Comment (domanchi) [inline tests/main_test.py]: Let's remove these `print` statements, before merging.

File: tests/pre_commit_hook_test.py [hunk 3]
```diff
@@ -44,17 +44,53 @@ def test_file_with_secrets(self, mock_log):
     def test_file_no_secrets(self):
         assert_commit_succeeds('test_data/files/file_with_no_secrets.py')
 
-    def test_baseline(self):
+    @pytest.mark.parametrize(
+        ' has_result, use_private_key_scan,, hook_command, commit_result',
```
Comment (domanchi) [inline tests/pre_commit_hook_test.py]: extra comma? maybe `s/commit_result/commit_succeeds`? (also prefixed space, but less fussy)

File: tests/main_test.py [hunk 4]
```diff
@@ -163,10 +164,186 @@ def test_old_baseline_ignored_with_update_flag(
...
+                    {
+                        "base64_limit": 3.5,
+                        "name": "Base64HighEntropyString",
+                    },
+                    {
+                        "name": "PrivateKeyDetector",
+                    },
+                ],
+            ),
+            (  # ignore overwriten option from CLI when not using --use-all-plugins
```
Comment (domanchi) [inline tests/main_test.py]: Hmm. If you **explicitly** override it by providing the CLI argument, even without `--use-all-plugins`, wouldn't you expect CLI to take precedence over a configured baseline?

Example use case: user wants to see if a decrease in sensitivity results in a large drop in noise.

```
$ detect-secrets scan --update .secrets.baseline --base64-limit 5
$ git diff .secrets.baseline
```

File: detect_secrets/plugins/common/initialize.py [hunk 5]
```diff
@@ -30,6 +30,31 @@ def from_parser_builder(plugins_dict):
...
+    if args.use_all_plugins:
+        plugins = from_parser_builder(args.plugins)
+    elif args.disabled_plugins:  # strip some plugins off baseline
+        plugins = _merge_plugin_from_baseline(baseline_plugins, args)
+    else:
+        plugins = baseline_plugins
+    return plugins
+
+
+def _merge_plugin_from_baseline(baseline_plugins, args):
```
Comment (KevinHock) [inline detect_secrets/plugins/common/initialize.py]: Maybe rename to `_trim_disabled_plugins_from_baseline`, it would make the make the comments in `merge_plugin_from_baseline` less needed.

File: detect_secrets/plugins/common/initialize.py [hunk 6]
```diff
@@ -30,6 +30,31 @@ def from_parser_builder(plugins_dict):
...
+    merged_plugins_dict = {vars(plugin)['name']: plugin for plugin in baseline_plugins}
```
Comment (KevinHock) [inline detect_secrets/plugins/common/initialize.py]: Great use of `vars` 👍

File: tests/main_test.py (hunk 4 above)
Comment (killuazhu) [reply tests/main_test.py]: I debated about that, it might have some edge cases. For example, if the baseline does not have the base64 scan, do we ignore the limit? or add the base64 scan in? So I chose to let the user on purposely use the plugin, then allow adjusting. How would you suggest to adjust?

File: tests/pre_commit_hook_test.py (hunk 3 above)
Comment (killuazhu) [reply tests/pre_commit_hook_test.py]: yep, extra comma. The rename sounds good. I will update in a new commit.

File: detect_secrets/plugins/common/initialize.py (hunk 6 above)
Comment (killuazhu) [reply detect_secrets/plugins/common/initialize.py]: Credit to @meneal

File: tests/main_test.py (hunk 4 above)
Comment (killuazhu) [reply tests/main_test.py]: It's updated to support these cases now. New test cases added.

**Comment by killuazhu** [comment]
@domanchi ready for another round.

File: detect_secrets/core/usage.py [hunk 7]
```diff
@@ -309,6 +309,7 @@ def consolidate_args(args):
 
         active_plugins = {}
         disabled_plugins = {}
```
Comment (domanchi) [inline detect_secrets/core/usage.py]: Now that I think about it, what **is** our purpose of returning a set of disabled_plugins? Since `args.plugins` is a collection of all active plugins (combining default values and specifically set values), wouldn't any plugin that isn't in this dictionary be disabled?

File: detect_secrets/core/usage.py [hunk 8]
```diff
@@ -309,6 +309,7 @@ def consolidate_args(args):
...
+        param_from_default = {}
```
Comment (domanchi) [inline detect_secrets/core/usage.py]: Maybe `is_default_parameter`, so it's clearer that it contains boolean values?

File: detect_secrets/core/usage.py [hunk 9]
```diff
@@ -309,6 +309,7 @@ def consolidate_args(args):
 
         active_plugins = {}
```
[the remark this answers is not in the document]
Comment (killuazhu) [reply detect_secrets/core/usage.py]: `disabled_plugins` is mainly used to filter off some disabled plugins when reading the plugin list from baseline. Relates to the comment below, it's possible to generate the `disabled_plugins` list on-demand by calculating the difference between all plugins and active plugins. I can add a helper function to the `usage.py`.

I was using `disabled_plugins` earlier to avoid re-iterating on the all plugins to get such list.

File: detect_secrets/core/usage.py (hunk 7 above)
Comment (killuazhu) [reply detect_secrets/core/usage.py]: See comment above.

File: detect_secrets/core/usage.py (hunk 8 above)
Comment (killuazhu) [reply detect_secrets/core/usage.py]: That's a good idea. I would make a suggestion to rename it to `is_using_default_value`

**Comment by killuazhu** [comment]
> Thanks for writing that hairy logic of plugin prioritization.

It definitely makes the job more challenge when I have to convert the BasePlugin tuple into a dictionary, then compares and sets values. I noticed the BasePlugin class has made `__dict__` field itself a property instead of the exact properties such as `name`, `base64_limit` and etc. This prevents the `{set|has|get}attr` from reading and writing values. May I ask what was the rationale of overwriting `__dict__` instead of directly using property on each field? @domanchi

File: detect_secrets/plugins/common/initialize.py [hunk 10]
```diff
@@ -32,28 +32,68 @@ def from_parser_builder(plugins_dict):
...
+        for param_name, param_value in list(plugin_params.items()):
+            from_default = args.param_from_default.get(param_name, False)
+            if from_default is False:
+                try:
+                    plugins_dict[plugin_name][param_name] = param_value
+                except KeyError:
+                    log.warning(
+                        '%s specified, but %s not configured! Ignoring...'
+                        % ("".join(["--", param_name.replace("_", "-")]), plugin_name),
+                    )
```
[the remark this answers is not in the document]
Comment (killuazhu) [reply detect_secrets/plugins/common/initialize.py]: It's been updated now. Although having more lines, but I do think the logic is a little bit more clear.

**Comment by killuazhu** [comment]
@domanchi all comments addressed.

File: detect_secrets/core/usage.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (domanchi) [reply detect_secrets/core/usage.py]: Gotcha. Makes sense.

Thanks for adding the helper function anyway!

File: detect_secrets/plugins/common/initialize.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (domanchi) [reply detect_secrets/plugins/common/initialize.py]: Clearer logic >> more lines, especially when it comes to long term maintenance 😸
